tools/tpm2_print_quote.c

```diff
@@ -0,0 +1,220 @@
+//**********************************************************************;
+// Copyright (c) 2017, National Instruments
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimer in the documentation
+// and/or other materials provided with the distribution.
+//
+// 3. Neither the name of Intel Corporation nor the names of its contributors
+// may be used to endorse or promote products derived from this software without
+// specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <stdio.h>
+#include "log.h"
+#include "tpm2_util.h"
+
+static UINT8 fdread_uint8(FILE* fd, bool* read_error) {
+    UINT8 x;
+    if(fread(&x, sizeof(UINT8), 1, fd) != 1) {
+        x = 0;
+        *read_error = true;
+    }
+    return x;
+}
+
+static UINT16 fdread_uint16(FILE* fd, bool* read_error) {
+    UINT16 x;
+    if(fread(&x, sizeof(UINT16), 1, fd) != 1) {
+        x = 0;
+        *read_error = true;
+    }
+    x = tpm2_util_ntoh_16(x);
+    return x;
+}
+
+static UINT32 fdread_uint32(FILE* fd, bool* read_error) {
+    UINT32 x;
+    if(fread(&x, sizeof(UINT32), 1, fd) != 1) {
+        x = 0;
+        *read_error = true;
+    }
+    x = tpm2_util_ntoh_32(x);
+    return x;
+}
+
+static UINT64 fdread_uint64(FILE* fd, bool* read_error) {
+    UINT64 x;
+    if(fread(&x, sizeof(UINT64), 1, fd) != 1) {
+        x = 0;
+        *read_error = true;
+    }
+    x = tpm2_util_ntoh_64(x);
+    return x;
+}
+
+static bool print_hex(FILE* fd, long unsigned int size, bool* read_error) {
```

Comment:
hex_dump exists in tpm2_nvread IIRC, move it to a common utility routine

Reply:
tpm2_nvread's hexdump() prints a multi-line output similar to the hexdump utility. I prefer a single-line output in tpm2_print so that it's easily parseable by grep/cut (in shell scripts). For example, I have a use case where we need to authenticate a quote and display only PCR digest and qualifier. This is trivial to implement in bash in the current form.